File: Components/DevExpressVCL/Demos/ExpressPivotGrid/CBuilder/CustomDrawDemo/CustomDrawMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "CustomDrawMain.h"
//---------------------------------------------------------------------------
#include "cxGraphics.hpp"

#pragma package(smart_init)
#pragma link "cxLookAndFeels"
#pragma link "DemoBasicMain"
#pragma link "cxControls"
#pragma link "cxCustomPivotGrid"
#pragma link "cxDBPivotGrid"
#pragma link "cxClasses"
#pragma link "cxCustomData"
#pragma link "cxStyles"
#pragma link "cxEdit"
#pragma link "cxLookAndFeelPainters"
#pragma resource "*.dfm"
TfrmCustomDraw *frmCustomDraw;

// we do not use dxOffice11.FillGradientRect because of the problem in C++Builder during linking
void __fastcall FillGradientRectEx(TcxCanvas* ACanvas, const TRect &ARect, TColor AColor1, TColor AColor2, bool AHorizontal)
{
  int AFirstOffset, ALastOffset, APixelSteps, AColorStepsR, AColorStepsG, AColorStepsB;
  TRect R = ARect;
  TColor AColor;
  if (AHorizontal) {
	AFirstOffset = ARect.Left;
	ALastOffset = ARect.Right - 1;
  }
  else
  {
    AFirstOffset = ARect.Top;
    ALastOffset = ARect.Bottom - 1;
  }
  APixelSteps = ALastOffset - AFirstOffset;
  AColorStepsR = GetRValue(AColor2) - GetRValue(AColor1);
  AColorStepsG = GetGValue(AColor2) - GetGValue(AColor1);
  AColorStepsB = GetBValue(AColor2) - GetBValue(AColor1);

  for (int I = AFirstOffset; I <= ALastOffset; I++)
  {
    if (AHorizontal) {
      R.Left = I;
	  R.Right = I + 1;
	}
    else
    {
      R.Top = I;
      R.Bottom = I + 1;
	}
	AColor = (TColor)RGB(
	  GetRValue(AColor1) + MulDiv(I - AFirstOffset, AColorStepsR, APixelSteps),
	  GetGValue(AColor1) + MulDiv(I - AFirstOffset, AColorStepsG, APixelSteps),
	  GetBValue(AColor1) + MulDiv(I - AFirstOffset, AColorStepsB, APixelSteps));
	ACanvas->FillRect(R, AColor);
  }
}


//---------------------------------------------------------------------------
__fastcall TfrmCustomDraw::TfrmCustomDraw(TComponent* Owner)
        : TfrmDemoBasicMain(Owner)
{
}

TcxLookAndFeelKind __fastcall TfrmCustomDraw::GetDefaultLookAndFeelKind()
{
  return(lfUltraFlat);
}

TcxCustomPivotGrid* __fastcall TfrmCustomDraw::PivotGrid()
{
  return DBPivotGrid;
}

//---------------------------------------------------------------------------

void __fastcall TfrmCustomDraw::DrawingClick(TObject *Sender)
{
  switch (((TMenuItem*)Sender)->Tag){
    case 1: {
	  if (((TMenuItem*)Sender)->Checked) {
        DBPivotGrid->OnCustomDrawCell = DBPivotGridCustomDrawCell;
      }
	  else {
        DBPivotGrid->OnCustomDrawCell = NULL;
      }
    break;
    }
    case 3: {
      if (((TMenuItem*)Sender)->Checked) {
        DBPivotGrid->OnCustomDrawColumnHeader = DBPivotGridCustomDrawColumnHeader;
        DBPivotGrid->OnCustomDrawRowHeader = DBPivotGridCustomDrawRowHeader;
      }
      else {
        DBPivotGrid->OnCustomDrawColumnHeader = NULL;
        DBPivotGrid->OnCustomDrawRowHeader = NULL;
      }
    }
  }
  PivotGrid()->LayoutChanged();
}
//---------------------------------------------------------------------------


void __fastcall TfrmCustomDraw::DBPivotGridCustomDrawColumnHeader(
	  TcxCustomPivotGrid *Sender, TcxCanvas *ACanvas,
      TcxPivotGridHeaderCellViewInfo *AViewInfo, bool &ADone)
{
  TPoint P = Sender->ViewData->FocusedCell;
  int AMin = (((TcxPivotGridViewDataItemAccess*)((TcxPivotGridHeaderCellViewInfoAccess*)AViewInfo)->Data))->GetChildLeftVisibleIndex();
  int AMax = (((TcxPivotGridViewDataItemAccess*)((TcxPivotGridHeaderCellViewInfoAccess*)AViewInfo)->Data))->GetChildRightVisibleIndex();
  if ((P.x >= AMin) && (P.x <= AMax))
	FSupportFont->Style = TFontStyles() << fsBold;
  else
  	FSupportFont->Style = TFontStyles();

  TcxViewParams AViewParams = AViewInfo->ViewParams;
  AViewParams.Font = FSupportFont;
  AViewInfo->ViewParams = AViewParams;
}
//---------------------------------------------------------------------------

void __fastcall TfrmCustomDraw::DBPivotGridCustomDrawRowHeader(
	  TcxCustomPivotGrid *Sender, TcxCanvas *ACanvas,
	  TcxPivotGridHeaderCellViewInfo *AViewInfo, bool &ADone)
{
  TPoint P = Sender->ViewData->FocusedCell;
  int AMin = (((TcxPivotGridViewDataItemAccess*)((TcxPivotGridHeaderCellViewInfoAccess*)AViewInfo)->Data))->GetChildLeftVisibleIndex();
  int AMax = (((TcxPivotGridViewDataItemAccess*)((TcxPivotGridHeaderCellViewInfoAccess*)AViewInfo)->Data))->GetChildRightVisibleIndex();
  if ((P.y >= AMin) && (P.y <= AMax))
	FSupportFont->Style = TFontStyles() << fsBold;
  else
	FSupportFont->Style = TFontStyles();

  TcxViewParams AViewParams = AViewInfo->ViewParams;
  AViewParams.Font = FSupportFont;
  AViewInfo->ViewParams = AViewParams;
}
//---------------------------------------------------------------------------

void __fastcall TfrmCustomDraw::DBPivotGridCustomDrawCell(
      TcxCustomPivotGrid *Sender, TcxCanvas *ACanvas,
      TcxPivotGridDataCellViewInfo *AViewInfo, bool &ADone)
{
  if (AViewInfo->LimitValueTypes.Contains(lvtColumnMaximum) && miLimitValues->Checked)
	ACanvas->Font->Color = clRed;
  else
	if (AViewInfo->LimitValueTypes.Contains(lvtColumnMinimum) && miLimitValues->Checked)
	  ACanvas->Font->Color = clLime;
	else
	{
	  int ACellPosition = AViewInfo->ColumnIndex + (AViewInfo->RowIndex << 16);
	  if (FCurrentColorIndex == -1)
	  {
		if ((FList->Count < 5) && (RandomRange(0, 20) == 1))
		  FList->Add(ACellPosition);
	  }
	  else
		if (FList->IndexOf(Pointer(ACellPosition)) > -1)
		{
		  ACanvas->Font->Style = TFontStyles() << fsBold;
		  ACanvas->Font->Color = FColors[FCurrentColorIndex];
		}
	}
  if (AViewInfo->RowIndex % 2 == 0)
	ACanvas->Brush->Color = clSilver;
}
//---------------------------------------------------------------------------
void __fastcall TfrmCustomDraw::DBPivotGridSelectionChanged(TObject *Sender)
{
  PivotGrid()->Invalidate();
}
//---------------------------------------------------------------------------
void __fastcall TfrmCustomDraw::FormCreate(TObject *Sender)
{
  PivotGrid()->DoubleBuffered = true;
  FSupportFont = new TFont();
  FList = new TcxPivotGridRecords();
  FCurrentColorIndex = -1;
  BYTE R = GetRValue(ColorToRGB(clWindowText));
  BYTE G = GetGValue(ColorToRGB(clWindowText));
  BYTE B = GetBValue(ColorToRGB(clWindowText));
  int I = 0;
  while (R < 255)
  {
	FColors[I] = (TColor)RGB(R, G, B);
	R = Min(R + 50, 255);
	I++;
  }
  int ACurrent = I;
  FColorCount = I * 2 - 1;
  for (I = ACurrent; I < FColorCount; I++)
	FColors[I] = FColors[FColorCount - 1 - I];
}
//---------------------------------------------------------------------------
void __fastcall TfrmCustomDraw::FormDestroy(TObject *Sender)
{
  delete FList;
  delete FSupportFont;
}
//---------------------------------------------------------------------------
void __fastcall TfrmCustomDraw::tmrColorChangeTimer(TObject *Sender)
{
  tmrColorChange->Enabled = false;
  FCurrentColorIndex++;
  if (FCurrentColorIndex >= FColorCount)
	FCurrentColorIndex = -1;
  if (FCurrentColorIndex == -1)
  {
	FList->Clear();
	tmrColorChange->Interval = 1000;
  }
  else
	tmrColorChange->Interval = 200;
  PivotGrid()->Invalidate();
  tmrColorChange->Enabled = true;
}
//---------------------------------------------------------------------------

File: Components/DevExpressVCL/Demos/ExpressPivotGrid/Delphi/CustomDrawDemo/CustomDrawMain.pas
unit CustomDrawMain;

{$I cxVer.inc}

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms,
  Dialogs, DemoBasicMain, cxLookAndFeels, Menus, StdCtrls, cxControls,
  cxCustomPivotGrid, cxDBPivotGrid, DemoBasicDM, cxGraphics, dxOffice11,
  cxClasses, cxCustomData, cxStyles, cxLookAndFeelPainters, cxEdit,
  ExtCtrls;

type
  TfrmCustomDraw = class(TfrmDemoBasicMain)
    DBPivotGrid: TcxDBPivotGrid;
    pgfPaymentType: TcxDBPivotGridField;
    pgfQuantity: TcxDBPivotGridField;
    pgfCarName: TcxDBPivotGridField;
    pgfUnitPrice: TcxDBPivotGridField;
    pgfCompanyName: TcxDBPivotGridField;
    pgfPaymentAmount: TcxDBPivotGridField;
    Drawing1: TMenuItem;
    Content1: TMenuItem;
    GroupHeaders1: TMenuItem;
    N3: TMenuItem;
    miLimitValues: TMenuItem;
    tmrColorChange: TTimer;
    procedure DBPivotGridCustomDrawCell(Sender: TcxCustomPivotGrid;
      ACanvas: TcxCanvas; AViewInfo: TcxPivotGridDataCellViewInfo;
      var ADone: Boolean);
    procedure DBPivotGridCustomDrawColumnHeader(Sender: TcxCustomPivotGrid;
      ACanvas: TcxCanvas; AViewInfo: TcxPivotGridHeaderCellViewInfo;
      var ADone: Boolean);
    procedure DBPivotGridCustomDrawRowHeader(Sender: TcxCustomPivotGrid;
      ACanvas: TcxCanvas; AViewInfo: TcxPivotGridHeaderCellViewInfo;
      var ADone: Boolean);
    procedure DrawingClick(Sender: TObject);
    procedure DBPivotGridSelectionChanged(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure tmrColorChangeTimer(Sender: TObject);
  private
    FCurrentColorIndex: Integer;
    FColors: array of TColor;
    FList: TcxPivotGridRecords;
    FSupportFont: TFont;
  protected
    function GetDefaultLookAndFeelKind: TcxLookAndFeelKind; override;
    function GetPivotGrid: TcxCustomPivotGrid; override;
  public
    { Public declarations }
  end;

var
  frmCustomDraw: TfrmCustomDraw;

implementation

{$R *.dfm}

uses
  DemoUtils, Math, Contnrs;

const
  cxSelectedFontStylesMap: array[Boolean] of TFontStyles = ([], [fsBold]);

type
  TcxPivotGridViewDataItemAccess = class(TcxPivotGridViewDataItem);
  TcxPivotGridHeaderCellViewInfoAccess = class(TcxPivotGridHeaderCellViewInfo);
  TcxCustomPivotGridAccess = class(TcxCustomPivotGrid);

function TfrmCustomDraw.GetDefaultLookAndFeelKind: TcxLookAndFeelKind;
begin
  Result := lfUltraFlat;
end;

function TfrmCustomDraw.GetPivotGrid: TcxCustomPivotGrid;
begin
  Result := DBPivotGrid;
end;

procedure TfrmCustomDraw.DBPivotGridCustomDrawCell(
  Sender: TcxCustomPivotGrid; ACanvas: TcxCanvas;
  AViewInfo: TcxPivotGridDataCellViewInfo; var ADone: Boolean);

  function GetCellPosition: Integer;
  begin
    Result := MakeLong(AViewInfo.ColumnIndex, AViewInfo.RowIndex);
  end;

begin
  if (lvtColumnMaximum in AViewInfo.LimitValueTypes) and dxDemoIsMenuChecked(dxDemoFindMenuItem(mmMain, 'miLimitValues')) then
    ACanvas.Font.Color := clRed
  else
    if (lvtColumnMinimum in AViewInfo.LimitValueTypes) and dxDemoIsMenuChecked(dxDemoFindMenuItem(mmMain, 'miLimitValues')) then
      ACanvas.Font.Color := clLime
    else
    begin
      if FCurrentColorIndex = -1 then
      begin
        if (FList.Count < 5) and (RandomRange(0, 20) = 1) then
          FList.Add(GetCellPosition);
      end
      else
        if FList.IndexOf(Pointer(GetCellPosition)) > -1 then
        begin
          ACanvas.Font.Style := [fsBold];
          ACanvas.Font.Color := FColors[FCurrentColorIndex];
        end;
    end;
  if (AViewInfo.RowIndex mod 2 = 0) or (AViewInfo.RowIndex mod 2 = 0) then
    ACanvas.Brush.Color := clSilver;
end;

procedure TfrmCustomDraw.DBPivotGridCustomDrawColumnHeader(
  Sender: TcxCustomPivotGrid; ACanvas: TcxCanvas;
  AViewInfo: TcxPivotGridHeaderCellViewInfo; var ADone: Boolean);
var
  P: TPoint;
  AMin, AMax: Integer;
  AViewParams: TcxViewParams;
begin
  P := Sender.ViewData.FocusedCell;
  with TcxPivotGridViewDataItemAccess(TcxPivotGridHeaderCellViewInfoAccess(AViewInfo).Data) do
  begin
    AMin := GetChildLeftVisibleIndex;
    AMax := GetChildRightVisibleIndex;
  end;
  FSupportFont.Style := cxSelectedFontStylesMap[(P.X >= AMin) and (P.X <= AMax)];
  AViewParams := AViewInfo.ViewParams;
  AViewParams.Font := FSupportFont;
  AViewInfo.ViewParams := AViewParams;
end;

procedure TfrmCustomDraw.DBPivotGridCustomDrawRowHeader(
  Sender: TcxCustomPivotGrid; ACanvas: TcxCanvas;
  AViewInfo: TcxPivotGridHeaderCellViewInfo; var ADone: Boolean);
var
  P: TPoint;
  AMin, AMax: Integer;
  AViewParams: TcxViewParams;
begin
  P := Sender.ViewData.FocusedCell;
  with TcxPivotGridViewDataItemAccess(TcxPivotGridHeaderCellViewInfoAccess(AViewInfo).Data) do
  begin
    AMin := GetChildLeftVisibleIndex;
    AMax := GetChildRightVisibleIndex;
  end;
  FSupportFont.Style := cxSelectedFontStylesMap[(P.Y >= AMin) and (P.Y <= AMax)];
  AViewParams := AViewInfo.ViewParams;
  AViewParams.Font := FSupportFont;
  AViewInfo.ViewParams := AViewParams;
end;

procedure TfrmCustomDraw.DrawingClick(Sender: TObject);
var
  AChecked: Boolean;
begin
  AChecked := dxDemoIsMenuChecked(Sender);
  case TComponent(Sender).Tag of
    1:
      if AChecked then
      begin
        DBPivotGrid.OnCustomDrawCell := DBPivotGridCustomDrawCell;
        dxDemoMenuItemSetChecked(dxDemoFindMenuItem(mmMain, 'miLimitValues'), True);
        dxDemoMenuItemSetEnabled(dxDemoFindMenuItem(mmMain, 'miLimitValues'), True);
      end
      else
      begin
        dxDemoMenuItemSetChecked(dxDemoFindMenuItem(mmMain, 'miLimitValues'), False);
        dxDemoMenuItemSetEnabled(dxDemoFindMenuItem(mmMain, 'miLimitValues'), False);
        DBPivotGrid.OnCustomDrawCell := nil;
      end;
    3:
      if AChecked then
      begin
        DBPivotGrid.OnCustomDrawColumnHeader := DBPivotGridCustomDrawColumnHeader;
        DBPivotGrid.OnCustomDrawRowHeader := DBPivotGridCustomDrawRowHeader;
      end
      else
      begin
        DBPivotGrid.OnCustomDrawColumnHeader := nil;
        DBPivotGrid.OnCustomDrawRowHeader := nil;
      end;
  end;
  PivotGrid.LayoutChanged;
end;

procedure TfrmCustomDraw.DBPivotGridSelectionChanged(Sender: TObject);
begin
  inherited;
  PivotGrid.Invalidate;
end;

procedure TfrmCustomDraw.FormCreate(Sender: TObject);
var
  I: Integer;
  R, G, B: Byte;
  ACount: Integer;
  ACurrent: Integer;
begin
  inherited;
  FSupportFont := TFont.Create;
  FList := TcxPivotGridRecords.Create;
  FCurrentColorIndex := -1;
  R := GetRValue(ColorToRGB(clWindowText));
  G := GetGValue(ColorToRGB(clWindowText));
  B := GetBValue(ColorToRGB(clWindowText));
  I := 0;
  while R < 255 do
  begin
    SetLength(FColors, I + 1);
    FColors[I] := RGB(R, G, B);
    R := Min(R + 50, 255);
    Inc(I);
  end;
  ACurrent := I;
  ACount := Length(FColors) * 2 - 1;
  SetLength(FColors, ACount);
  for I := ACurrent to ACount - 1 do
    FColors[I] := FColors[ACount - 1 - I];
  PivotGrid.DoubleBuffered := True;
end;

procedure TfrmCustomDraw.FormDestroy(Sender: TObject);
begin
  SetLength(FColors, 0);
  FreeAndNil(FList);
  FreeAndNil(FSupportFont);
  inherited;
end;

procedure TfrmCustomDraw.tmrColorChangeTimer(Sender: TObject);
begin
  tmrColorChange.Enabled := False;
  Inc(FCurrentColorIndex);
  if FCurrentColorIndex >= Length(FColors) then
    FCurrentColorIndex := -1;
  if FCurrentColorIndex = -1 then
  begin
    FList.Clear;
    tmrColorChange.Interval := 1000;
  end
  else
    tmrColorChange.Interval := 200;
  PivotGrid.Invalidate;
  tmrColorChange.Enabled := True;
end;

end.
